src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -9,58 +9,131 @@ namespace System.Numerics
 {
     internal static partial class BigIntegerCalculator
     {
+
+#if DEBUG
+        // Mutable for unit testing.
+        private static
+#else
+        private const
+#endif
+        int CopyToThreshold = 8;
+
+        private static void CopyTail(ReadOnlySpan<uint> source, Span<uint> dest, int start)
+        {
+            source.Slice(start).CopyTo(dest.Slice(start));
+        }
+
         public static void Add(ReadOnlySpan<uint> left, uint right, Span<uint> bits)
         {
             Debug.Assert(left.Length >= 1);
             Debug.Assert(bits.Length == left.Length + 1);
 
+            // Switching to managed references helps eliminating
+            // index bounds check...
+            ref uint resultPtr = ref MemoryMarshal.GetReference(bits);
+            ref uint leftPtr = ref MemoryMarshal.GetReference(left);
+
             // Executes the addition for one big and one 32-bit integer.
             // Thus, we've similar code than below, but there is no loop for
             // processing the 32-bit integer, since it's a single element.
 
+            nint i = 0;
+            nint upperBound = left.Length;
             long carry = right;
 
-            for (int i = 0; i < left.Length; i++)
+            if (upperBound <= CopyToThreshold)
             {
-                long digit = left[i] + carry;
-                bits[i] = unchecked((uint)digit);
-                carry = digit >> 32;
+                for ( ; i < upperBound; i++)
+                {
+                    carry += Unsafe.Add(ref leftPtr, i);
+                    Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                    carry >>= 32;
+                }
+
+                Unsafe.Add(ref resultPtr, upperBound) = unchecked((uint)carry);
+            }
+            else
+            {
+                for ( ; i < upperBound; )
+                {
+                    carry += Unsafe.Add(ref leftPtr, i);
+                    Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                    i++;
+                    carry >>= 32;
+                    if (carry == 0) break;
```

Comment:
it would be nice to have a comment explaining what it does here. I've just copy-pasted your own explanation from https://github.com/dotnet/runtime/issues/83457#issue-1625669516 and added a bracket for style consistency.

```suggestion
                    // Once carry is set to 0 it can not be 1 anymore. 
                    // So the tail of the loop is just the movement of argument values to result span.
                    if (carry == 0)
                    {
                        break;
                    }
```